torch/_ops.py

```diff
@@ -228,22 +287,44 @@ def inner(fn):
 
         return inner
 
-    def dispatch(self, dispatch_key, *args, **kwargs):
-        if dispatch_key == torch._C.DispatchKey.Python:  # type: ignore[attr-defined]
-            # TODO(voz): We should walk all the nodes here / turn it into a list, topmode is ok for now.
-            curr_mode = type(torch._C._get_torch_dispatch_mode())
-            assert (
-                curr_mode is not None
-            ), "Illegal invocation of dispatch on torch._C.DispatchKey.Python without a mode."
-            if curr_mode not in self.python_key_mode_table:
-                return self._op_dk(dispatch_key, *args, **kwargs)
-            # TODO(voz): The idea behind this is that we do not yet support dispatch by key + mode, only key.
-            return self.python_key_mode_table[curr_mode](*args, **kwargs)
+    def __getattr__(self, attr):
+        if len(attr) == 0 or not attr[0].isupper():
+            raise AttributeError()
 
-        if dispatch_key in self.py_kernels:
-            return self.py_kernels[dispatch_key](*args, **kwargs)
-        else:
-            return self._op_dk(dispatch_key, *args, **kwargs)
+        try:
+            key = torch._C._dispatch_key_parse(attr)
+        except Exception as e:
+            import traceback
+
+            traceback.print_exc()
+            raise AttributeError()
+
+        if key == torch._C.DispatchKey.Python:  # type: ignore[attr-defined]
+            if not self.python_key_mode_table:
+                setattr(self, attr, key)
+                return key
+
+            def handler(*args, **kwargs):
+                # TODO: We also need to handle tensor subclasses here
+                # TODO(voz): We should walk all the nodes here / turn it into a list, topmode is ok for now.
+                curr_mode = type(torch._C._get_torch_dispatch_mode())
+                assert (
+                    curr_mode is not None
+                ), "Illegal invocation of dispatch on torch._C.DispatchKey.Python without a mode."
+                if curr_mode not in self.python_key_mode_table:
+                    # TODO: This path is slow, should generally encourage this
+                    # case to not happen
+                    return self._op_dk(key, *args, **kwargs)
+                # TODO(voz): The idea behind this is that we do not yet support dispatch by key + mode, only key.
+                return self.python_key_mode_table[curr_mode](*args, **kwargs)
+
+            setattr(self, attr, handler)
+            return handler
```

Comment:
what's the deal with returning a handler function instead of a DispatchKey object?  i'm trying to find where this getattr is used and how it is determined whether to call its return value

Reply:
If you return a DispatchKey object, we directly call the c++ kernel for the dispatch key. If you return a function, we call the function to handle the kernel. The getattr is triggered by https://github.com/pytorch/pytorch/pull/85133/files/d5013fcde2ff0a9d9587614f77da73b4b934d331#diff-0fa862a86ab690b67dbc2aecb5cc0675d6df08af87e8a2b0655e554cef6f11d7

Reply:
could you add that in a comment somewhere in `def __getattr__`? (that the contract is we return either a python handler fn, or a dispatch key). The caller-callee stuff is very non-local :)